### PR TITLE
Fix GraphQL Errors When Updating Entity Relations fixes #1993

### LINKED ISSUE #1993: Look for the server-side GQL error
When updating datetime by only changing the related tickets or vice versa, it results in an unknown error. I think the error is caused because there are no changes in the actual entity, rather only the related items change. The source of error can be found by dumping he result of `$result = $entity->save($args);` in `DatetimeUpdate` and `TicketUpdate` under `EventEspresso\core\domain\services\graphql\mutators`.

### PR DESCRIPTION
**Pull request by tn3rb**
plz see #1993

There were a couple of issues happening that resulted in errors appearing in the GQL responses:

 - `DatetimeUpdate`, `TicketUpdate`, and `PriceUpdate` we were using the return value from `EE_Base_Class::save()` to determine whether the update was successful, but entities don't save if there were no changes made to them and that method returns `0`. The solution was simply to wrap the calls to `save()` in a try catch block and handle any actual errors that occur.

 - in `useOnSubmitPrices` values were getting submitted that were not of the correct type, such as `"false"` (string)  instead of `false` (boolean), so the solution was to normalize any values we know could cause problems.

- also fixed a couple other places where values were being set to ensure they were the correct data type

- also fixed another hook related problem in the ticket price calculator

### CODE REVIEW

File: assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js [hunk 1]
```diff
@@ -31,11 +31,20 @@ const useOnSubmitPrices = (existingPrices) => {
...
 			}
 			const { id, ...input } = pick(PRICE_INPUT_FIELDS, price);
+
+			const normalizedPriceFields = {
+				...input,
+				amount: parseFloat(price.amount || 0),
+				isDefault: !! price.isDefault,
+				order: parseInt(price.order, 10),
+				overrides: parseInt(price.overrides, 10),
+				parent: parseInt(price.parent, 10),
```
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js]: Many of these fields may not be present in `input`. Also, `parent` is `id` (GUID) of another price.

File: assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js [hunk 2]
```diff
@@ -47,8 +56,13 @@ const useOnSubmitPrices = (existingPrices) => {
 			deleteEntity({ id });
 		});
 
+		const normalizedTicketFields = {
+			...ticket,
+			price: parseFloat(ticket.price || 0),
+			reverseCalculate: !! ticket.reverseCalculate,
```
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js]: `ticket.reverseCalculate` is observed to be a string e.g. `"false"`, `!! ticket.reverseCalculate` will always be true.

File: assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useTicketPriceCalculator.js [hunk 3]
```diff
@@ -1,18 +1,27 @@
...
-		default:
-			return state;
-	}
-};
+const useTicketPriceCalculator = () => useCallback(
+	(state, action) => {
+		switch (action.type) {
+			case 'CALCULATE_BASE_PRICE':
+				return calculateBasePrice(state);
+			case 'CALCULATE_TICKET_TOTAL/':
```
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useTicketPriceCalculator.js]: May be that `/` is a typo?

File: core/domain/services/graphql/mutators/DatetimeUpdate.php [hunk 4]
```diff
@@ -83,13 +87,24 @@ public static function mutateAndGetPayload(EEM_Datetime $model, Datetime $type)
...
+                $entity->save($args);
+            } catch (Exception $exception) {
+                new ExceptionStackTraceDisplay(
+                    new RuntimeException(
+                        sprintf(
+                            esc_html__(
+                                'The Datetime failed to update because of the following error(s):%1$s%2$s',
+                                'event_espresso'
+                            ),
+                            '<br/>',
```
Comment (manzoorwanijk) [inline core/domain/services/graphql/mutators/DatetimeUpdate.php]: GraphQL clients usually treat those messages as escaped text, HTML is no rendered.

File: core/domain/services/graphql/mutators/PriceUpdate.php [hunk 5]
```diff
@@ -76,10 +79,21 @@ public static function mutateAndGetPayload(EEM_Price $model, Price $type)
...
+                $entity->save($args);
+            } catch (Exception $exception) {
+                new ExceptionStackTraceDisplay(
+                    new RuntimeException(
+                        sprintf(
+                            esc_html__(
+                                'The Price failed to update because of the following error(s):%1$s%2$s',
+                                'event_espresso'
+                            ),
+                            '<br/>',
```
Comment (manzoorwanijk) [inline core/domain/services/graphql/mutators/PriceUpdate.php]: Same here

File: assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js (hunk 2 above)
Comment (nicolad) [reply assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js]: Let's avoid using double negation trick.
More info on why:
- https://www.sitepoint.com/javascript-double-negation-trick-trouble/
- https://riptutorial.com/javascript/example/3047/double-negation----x-

File: assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js (hunk 1 above)
Comment (tn3rb) [reply assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useOnSubmitPrices.js]: oh right... forgot those are guids client side

File: assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useTicketPriceCalculator.js (hunk 3 above)
Comment (tn3rb) [reply assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useTicketPriceCalculator.js]: DOH! yes it is...  thnx

File: assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useTicketPriceCalculatorFormDecorator.js [hunk 6]
```diff
@@ -42,15 +42,15 @@ const useTicketPriceCalculatorFormDecorator = () => {
...
 			const result = calculator(formData, { type: 'CALCULATE_BASE_PRICE' });
 			const newBasePrice = getValue('prices[0].amount', result);
-			return { ['prices[0].amount']: newBasePrice };
+			return { ['prices[0].amount']: parseFloat(newBasePrice || 0) };
 		},
 		[calculator]
 	);
 	const updateTicketTotal = useCallback(
 		(formData) => {
 			const result = calculator(formData, { type: 'CALCULATE_TICKET_TOTAL/' });
```
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useTicketPriceCalculatorFormDecorator.js]: `/` here a well

Comment (tn3rb) [reply assets/prototype/domain/eventEditor/ticketsList/ticketPriceCalculator/hooks/useTicketPriceCalculatorFormDecorator.js]: OH DANG!

**Pull request merged**